[RFC 140] Simple package paths, part 1b: Enabling the directory structure

infinisil opened the pull request:
> **Note**
> This was reviewed by and merged on 2023-09-05 15:30 CEST with the [Nixpkgs Architecture Team](https://nixos.org/community/teams/nixpkgs-architecture.html) and others that joined the call!

### Context

This is the [first part](https://github.com/NixOS/rfcs/blob/master/rfcs/0140-simple-package-paths.md#pr-1-the-directory-structure) of the implementation of the accepted [RFC 140](https://github.com/NixOS/rfcs/pull/140), the introduction of the new directory structure. Note that this was split up into Part 1a (https://github.com/NixOS/nixpkgs/pull/250885) and Part 1b (this PR). The second part will be done afterwards, it's draft can be found here: https://github.com/NixOS/nixpkgs/pull/211832. This is hopefully only the first of many quality-of-life improvements to Nixpkgs by the [Nixpkgs Architecture Team](https://nixos.org/community/teams/nixpkgs-architecture.html).

### Overview

This RFC introduces the new `pkgs/by-name` directory, which makes it possible to add a new top-level package by just declaring it in `pkgs/by-name/${prefix}/${name}/package.nix`, where `name` is the attribute name the package should be exposed as, and `prefix` is the lowercased two-letter prefix of the attribute name. This means that for a lot of packages there's no more need to find an appropriate location for the file in the loosely categorized `pkgs/` hierarchy, and no more editing of the `all-packages.nix` file!

#### Restrictions
Note however that only a subset of packages may be using this new directory structure for now, extending this is future work for the Nixpkgs Architecture Team. In particular these requirements have to be met for an attribute to be defined in `pkgs/by-name`:
- It's at the top level, not allowed are packages from package sets like `pkgs.python3Packages.requests`.
- It's a derivation, not allowed are non-derivations like `pkgs.fetchFromGitHub`.
- It's defined using `callPackage`, not allowed are packages defined using functions like `python3Packages.buildPythonApplication`.

### Things done

- [x] Implement the automatic calling of packages in the `pkgs/by-name` directory and move `pkgs.hello` to it
  - [x] Working
  - [x] Optimised
  - [x] Tested
  - [x] Clean and documented code
- [x] Implement CI checks enforcing the structure and semantics of the `pkgs/by-name` directory 
  - [x] Working, you can run `nix-build -A tests.byName.nixpkgs`
  - [x] Is getting run by CI (see the checks in this PR)
    - [x] Invalid `pkgs/by-name` breaks CI ([example](https://github.com/NixOS/nixpkgs/actions/runs/6042323036/job/16397177271?pr=237439))
  - [x] Tested, you can run `nix-build -A tests.byName.tests`
  - [x] Clean and documented code
- [x] Update the contributor documentation for the `pkgs/by-name` directory and recommend using it for new packages
- [x] Add myself to relevant code owners entries

piegamesde commented on doc/contributing/coding-conventions.chapter.md:
```diff
@@ -230,185 +230,43 @@ There are a few naming guidelines:
+The `package.nix` file must be a function taking other top-level attributes or configuration flags as arguments, and returning a derivation.
+Arbitrary other files may be created in the directory if desired.
+Package files declared in this way are automatically called with `callPackage` and added to the top-level package set at `pkgs.$name`.
 
```
Maybe move the libfoo example up here, to interleave examples and description.

piegamesde commented on doc/contributing/coding-conventions.chapter.md:
````diff
@@ -230,185 +230,43 @@ There are a few naming guidelines:
+```
 
-**Else:**
+### Existing hierarchy {#sec-hierarchy}
````
I think it's fair to remove the current categorization stuff. What I'm missing here though is some information about when to use the old hierarchy

piegamesde commented on doc/contributing/quick-start.chapter.md:
```diff
@@ -50,28 +53,20 @@ To add a package to Nixpkgs:
+   To read more about the Nix language, see [Nix language basics](https://nix.dev/tutorials/first-steps/nix-language).
 
-5. To test whether the package builds, run the following command from the root of the nixpkgs source tree:
+4. To test whether the package builds, run the following command from the root of the nixpkgs source tree:
```
I've learned that you can always use `1.` in markdown and it will count correctly for you in the html. Looks a bit weird in code review, but reduces diff noise.

piegamesde commented on pkgs/top-level/stage.nix:
```diff
@@ -269,6 +269,12 @@ let
+  unitPackages = self: super:
+    lib.mapAttrs (name: file:
+      self.callPackage file { }
+    ) (import ./unit-files.nix);
```
Off-topic: this snippet made me think a lot about how to format things in nixpkgs, especially since it has a function call with long first argument and short last argument. I remember you saying about how to use `flip` in these cases, but actually I think that using `<|` before the last argument would work even better. Yet another reason why I want `<|` in the language …

piegamesde commented on pkgs/top-level/unit-files.nix:
```diff
@@ -0,0 +1,89 @@
+
+  # Return the name mappings in a single shard
+  # Type: String -> AttrsOf Path
+  namesForShard = shard:
```
If this function returned a list of name value pairs instead, and calling `builtins.listToAttrs` only after merging all lists, what would be the performance characteristic of that? Would it allow us to get rid of the binaryMerge? Especially given that `builtins.concatLists` exists, but maybe I'm naive to hope that it would be faster because of that.

infinisil replied:
Lists unfortunately have the same performance characteristics as attribute sets: `++` has complexity $O(l + r)$ with $l$ and $r$ being the sizes of the left and right side respectively. So because you'd still need to `foldl'` the lists using `++`, it would give poorer performance again.

infinisil replied:
Oh right, that might work! I'll give it a go when I work on this the next time!

infinisil commented on pkgs/top-level/unit-files.nix:
```diff
@@ -0,0 +1,105 @@
+  #
+  # This way the shard merging can be done by Nix's C++ code.
+
+  # We benchmark the three approaches with the following command with 300 runs:
+  #
+  #   NIX_SHOW_STATS=1 nix-instantiate --eval --strict 'builtins.seq (import pkgs/top-level/unit-files.nix) null'
+  # Approach     | cpuTime [ms] | gc.totalBytes [MB]
+  # linear merge | 33.4 ± 4.8   | 126.075
+  # concatMap    | 21.9 ± 2.6   |   5.824
+  # binary merge | 16.6 ± 2.5   |   7.102
+  #
+  # The lowest evaluation time is achieved with the binary merge algorithm, even despite the slightly higher memory usage.
+  # However if https://github.com/NixOS/nix/issues/7676 were implemented in the future, the concatMap-based approach could be much more performant.
+  # Alternatively, other Nix builtins to speed up this computation would help too.
```
@piegamesde Turns out it's a bit slower

piegamesde replied:
Unfortunate. We'd need a builtin like `concatAttrs` and maybe also `concatMapAttrs` for this then.

What about exposing your `binaryMerge` as `lib.concatAttrs`, so that a) people don't have to ad-hoc define it like done here and b) there is a greater chance of elevating this into a builtin later on?

infinisil replied:
The interface that would have to be exposed to get the same efficiency is `concatAttrs :: [String] -> (String -> AttrsOf a) -> AttrsOf a`, but the arguably expected interface would instead be `concatAttrs :: [AttrsOf a] -> AttrsOf a`. Because of this I'd prefer leaving it out of the discussion for now.

roberth replied:
`concatMapAttrs` being a `foldl'`, it is very close to the _linear merge_ implementation, so I would not expect it to perform better as implemented. Turning it into an optimized primitive in nix itself may be worth it though.

> I'm unhappy about that name [`concatMapAttrs`]

I am to blame for that, see discussion in
 - https://github.com/NixOS/nixpkgs/pull/201527

I'd be open to discuss it there if you can think of a better name that follows a useful naming convention. It's less than a year old, so I think it's worth renaming _if_ it produces a significant improvement.

roberth replied:
Fwiw Nix, as implemented in `NixOS/nix`, can't make use of the property that the attributes are already in lexicographical order. The internal ordering of attributes is an arbitrary ordering on internalized strings. Nonetheless, I expect a native implementation to be able to beat what we can do in the language.

piegamesde replied:
Oh, I didn't know `concatMapAttrs` was actually a thing in `lib`, the one I was describing was some made-up thing with a made-up name. The difference between the two ist, that the one in nixpkgs takes an attribute set as input while what we currently need is a list as input.

infinisil replied:
> Interesting, so `concatAttrs (map namesForShard shards)` would have worse performance then?

It does additionally have to allocate a ~700 item list, though this probably doesn't have a significant performance decrease.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-06-13-nixpkgs-architecture-team-meeting-40/29087/1

roberth commented on pkgs/top-level/stage.nix:
```diff
@@ -269,6 +269,12 @@ let
   };
 
+  # Auto-calls packages declared in the named package hierarchy in pkgs/unit
+  unitPackages = self: super:
+    lib.mapAttrs (name: file:
+      self.callPackage file { }
+    ) (import ./unit-files.nix);
```
```suggestion
  unitPackages = self: super:
    lib.mapAttrs
      (name: file: self.callPackage file { })
      (import ./unit-files.nix);
```

Motivation: arguments to the same curried function have the same indentation.

infinisil replied:
Interesting how you're suggesting this here, that's the exact style me and @piegamesde are going with for NixOS/rfcs#101! Will apply :smile:

RaitoBezarius commented on pkgs/top-level/unit-files.nix:
```diff
@@ -0,0 +1,105 @@
+  #     /\     O((n-1)m + m) = O(nm)
+  #    ⋰  m    ... (n times)
+  #   /\       O(2m + m)     = O(3m)
+  #  /\ m      O(m + m)      = O(2m)
```
nit(mathematician): `O(m)` :p

7c6f434c replied:
It's pretty clear that the point here is that the constants in O are «roughly» the same, and then we want the bottom to say 2m

piegamesde replied:
nit²: `O(m) = O(m + m)`, so so both are correct

7c6f434c replied:
Only as long as your O(2m), O(3m)… O(nm) do not want to have the same constant.

Ideally it would even be `Θ(·)`, not `O(·)`…

RaitoBezarius replied:
I mean, more precisely, summing $O(Km)$ with K constant is a bit clunky, this is a nit, but to the non-expert, it can give the impression that $\sum_k O(km) = O(m \sum_k k) = O(m(m + 1)/2) = O(m^2)$ I imagine.

I don't really care, just wanted to point out that replacing $O$ by a proper sequence $T(n, m)$ counting the number of elementary operations wrt to `//` might be enough, and you can conclude with any Landau notation that you prefer.

infinisil replied:
However, I believe it is the case that $\sum_{k=1}^{n} \mathcal{O}(km) \subseteq \mathcal{O(n^2m)}$, which is used here. Though I admit that originally I did it with the triangle formula before simplifying :)

infinisil replied:
Although, since the argument here is that the linear algorithm is slower, I'd have to prove that it's worse than the binary merge algorithm, so I believe to be fully correct I should calculate the $\Omega$ (a lower bound) for the linear algorithm and $\mathcal{O}$ (an upper bound) for the binary merge algorithm.

At which point just counting the operations like @RaitoBezarius suggested seems much easier!

RaitoBezarius replied:
No, consider the almost finite sequence u_n^k = kn if n <= k else 0.
All u_n^k are O(kn) where k is a fixed constant.
But sum of u_n^k is not an O(n^2) because you cannot bound uniformly the constant for all k

(Example written quickly, I am not totally sure it works but it sure feels a counterexample)

infinisil replied:
I see, thanks! I now changed it to count the number of copies directly, and only using $\mathcal{O}$ for the total

RaitoBezarius commented on pkgs/top-level/unit-files.nix:
```diff
@@ -0,0 +1,105 @@
+  #  /\   /\   /\   /\  O(n / 2 * (m + m))     = O(nm)
+  # m  m m  m m  m m  m
+  #
+  # Giving a total complexity of only O(log2(n)nm)
```
nit(mathematician): `O(mn log n)`

infinisil replied:
I'm using `log` instead of `log2` now, but I think I like the original order more because it's more directly comparable to the $n^2m$ result above

Artturin commented on .github/CODEOWNERS:
```diff
@@ -47,6 +47,11 @@
 
+# Unit directories
+/pkgs/top-level/unit-files.nix                   @infinisil @roberth
+/pkgs/unit/Readme.md                             @infinisil @roberth
```
```suggestion
/pkgs/unit/README.md                             @infinisil @roberth
```

for consistency

infinisil replied:
Done

2xsaiko commented on pkgs/test/named-hierarchy/reference/src/main.rs:
```diff
@@ -0,0 +1,147 @@
+    reference: PathBuf,
+    base_dir: &PathBuf,
+) -> Result<PathBuf> {
+    let mut rel_to_root = source.parent().unwrap().to_path_buf();
```
Won't work as is, PathBuf::parent returns Option and this function returns Result.

infinisil replied:
This condition can never occur because this function is only called with files, which always have a parent. Because of this I think `unwrap` is okay here, but I'll add a comment justifying why

infinisil commented on pkgs/test/named-hierarchy/reference/src/main.rs:
```diff
@@ -0,0 +1,147 @@
+            Component::CurDir => {}
+            Component::ParentDir => {
+                if !ascending {
+                    return Err(Error::msg(format!(
```
Nice, didn't know about this!

algorithmiker reviewed (commented):
I am not sure you knew you can interpolate format strings in rust :) 

Left diffs if you want to do it this way, but because it's mostly stylistic, I won't push it if you don't like it

infinisil commented:
@gilice I did not know that, neat! Greatly appreciated, I barely know Rust :D

Artturin commented on pkgs/test/named-hierarchy/reference/src/main.rs:
```diff
@@ -0,0 +1,139 @@
+        let text = node.text().to_string();
+        let line = line_index.line(node.text_range().start().into());
+
+        // Filters out ./foo/${bar}/baz
```
path interpolation isn't available in nix 2.3, a comment should be added that mentions that it can be removed after minver is 2.4+

infinisil commented on doc/contributing/file-structure.chapter.md:
I noticed that the overall file structure of Nixpkgs overall is not explained anywhere, and it's really hard to document the new structure without that, so I started a new section here where I document that structure.

I will expand this new section with more in-depth information

infinisil replied:
This was done with https://github.com/NixOS/nixpkgs/pull/245243, now it's here: https://github.com/NixOS/nixpkgs/blob/6d9caf9821e000de6fc846c90136705a7e966442/CONTRIBUTING.md#overview, and in `pkgs/README.md` :)

roberth commented on pkgs/top-level/named-hierarchy-files.nix:
```diff
@@ -0,0 +1,108 @@
+Type: AttrsOf Path
+*/
+
+# Because the result might be fairly large, we only want compute it once.
+# In order to ensure this, we make use of Nix's `import` cache.
+# This cache memoizes the resulting value of the imported file, notably including any `let in` bindings at the top of the file.
+# This is why everything here is computed in a `let in` binding at the top.
```
This could be moved to a test. The test is as simple as `isAttrs (import <this file>)`.

roberth commented on pkgs/top-level/named-hierarchy-files.nix:
```diff
@@ -0,0 +1,108 @@
+
+  # Merge the name mappings of all shards between the indices start (inclusive) and end (exclusive)
+  # This ignores any potential duplicates for efficiency
+  # Invariant: start < end
+  # Type: Int -> Int -> AttrsOf Path
+  binaryMerge = start: end:
```
```suggestion
  # Type: Int -> Int -> AttrsOf Path
  binaryMerge = start: end:
    # Invariant: start < end
```

Names aren't part of the public interface of a function. (attribute names would be)

infinisil replied:
Not sure what you mean to say with that, the suggestion to move the invariant into the function sounds okay though

roberth replied:
I guess I'm extrapolating from
 - name changes are not observable by callers
 - names used in a function implementation aren't always optimal, either for brevity or to avoid shadowing something in their scope

Combining those, you may conclude that argument names should not be considered as part of the public interface, although it is somewhat subjective.

I suppose you could use an `assert`, or a commented out one so we don't have to think too much about performance

```suggestion
  # Type: Int -> Int -> AttrsOf Path
  binaryMerge = start: end:
    # assert start < end;
```

infinisil replied:
Ohh I get it now, the problem is that the invariant uses variables available only inside the function. Makes sense, I'll apply this suggestion

roberth commented on pkgs/top-level/stage.nix:
```diff
@@ -269,6 +269,12 @@ let
   };
 
+  # Auto-calls packages declared in the named package hierarchy in pkgs/by-name
+  namedHierarchyPackages = self: super:
```
```suggestion
  autocalledPackages = self: super:
```

infinisil replied:
Now done in the latest push.

roberth commented on doc/contributing/file-structure.chapter.md:
```diff
@@ -0,0 +1,24 @@
+- `pkgs`: Entry-points, packages, builders
+  - `top-level`: Entrypoints, package set aggregations
+    - `impure.nix`, `default.nix`, `config.nix`: Root entry-point (`import <nixpkgs>`) function definition
+    - `stage.nix`, `all-packages.nix`, `named-hierarchy-files.nix`, `splice.nix`: Top-level attribute set (`import <nixpkgs> {…}`) aggregation definition
```
```suggestion
    - `stage.nix`, `all-packages.nix`, `named-hierarchy-files.nix`, `splice.nix`: Top-level attribute set (`import <nixpkgs> {…}`) aggregation definition also known as `pkgs`
```

infinisil replied:
This was part of https://github.com/NixOS/nixpkgs/pull/245243, though it was changed a bit: https://github.com/NixOS/nixpkgs/tree/b978241b83a7702dc2678d97dd03789073ca715c/pkgs#overview, I'll resolve this

roberth commented on pkgs/test/named-hierarchy/default.nix:
```diff
@@ -0,0 +1,184 @@
+  - bar
+    - bar (regular)
+  - fo
+    - @
```
What is this `@`?

infinisil replied:
This code in this file is still a bit drafty, but this is an example of an invalid package directory/attribute name. I will clean this code up

roberth commented on pkgs/test/named-hierarchy/default.nix:
```diff
@@ -0,0 +1,184 @@
+  - pkgs/by-name/fo: Duplicate case-sensitive entries:
+    - pkgs/by-name/fo/foo
+    - pkgs/by-name/fo/FOO
+  - pkgs/by-name/fo/@:
```
Oh, I thought you said is was part of the directory structure.

roberth commented on pkgs/test/named-hierarchy/default.nix:
```diff
@@ -0,0 +1,184 @@
+      throw ''
+        Semantic evaluation checks for pkgs/by-name failed:
+        ${lib.concatStrings semanticErrors}'';
+
```
That's a lot of untested code up to this point. How do we know that this tests what it's supposed to test? - Does it really catch mistakes?

I'd recommend to factor these functions out and give them tests.

Also, or alternatively, you could use [testBuildFailure](https://nixos.org/manual/nixpkgs/stable/#tester-testBuildFailure) to test the end result of the derivation below, if you make the `baseDirectory` a parameter, and then feed it some bad inputs that way.

infinisil replied:
Yes there's a TODO in the PR description to "Test the checks", will definitely do this!

infinisil replied:
I now added tests for all kinds of errors that could occur.

2xsaiko commented on doc/contributing/coding-conventions.chapter.md:
```diff
@@ -230,185 +230,43 @@ There are a few naming guidelines:
 Names of files and directories should be in lowercase, with dashes between words — not in camel case. For instance, it should be `all-packages.nix`, not `allPackages.nix` or `AllPackages.nix`.
 
-### Hierarchy {#sec-hierarchy}
+### New top-level package hierarchy {#sec-named-hierarchy}
```
"Name-based"*

infinisil commented on doc/contributing/file-structure.chapter.md:
```diff
@@ -0,0 +1,24 @@
+- `nixos`: NixOS, see the [Development section in the NixOS manual](https://nixos.org/nixos/manual/#ch-development)
+- `pkgs`: Entry-points, packages, builders
+  - `top-level`: Entrypoints, package set aggregations
+    - `impure.nix`, `default.nix`, `config.nix`: Root entry-point (`import <nixpkgs>`) function definition
```
Neither is `default.nix` nor `impure.nix` actually, only the `default.nix` in the Nixpkgs root is the actual entry-point. But the description says that it's not the entry-point itself, but rather the function definition of it, which these three files are involved with defining. Roughly, `impure.nix` for impure argument defaults, `default.nix` for the result of the function, `config.nix` to validate the `config` argument.

roberth replied:
`pkgs/top-level/default.nix` has been recommended as an entrypoint with fewer eval impurities, but perhaps only informally.

`config.nix` may be involved in the implementation of the entrypoint-function, but so is `stage.nix` or basically anything. So `config.nix` is not special in this regard. What is the difference between a dependency like `config.nix` or `lib.evalModules` or `stage.nix` which are equally necessary for the entrypoint to work? Without any of those, you don't get anything from the entrypoint.

infinisil replied:
Fair, but this is only a very rough categorization, I think it's accurate enough for the purposes here. Note that with this categorization, all files in `pkgs/top-level` are assigned to a category, I'd like to maintain this property

infinisil replied:
This was part of https://github.com/NixOS/nixpkgs/pull/245243: https://github.com/NixOS/nixpkgs/blob/9d303fba8af0ed985e33bb5ce7a84ced8a5f270f/pkgs/README.md#overview

infinisil commented on pkgs/top-level/named-hierarchy-files.nix:
```diff
@@ -0,0 +1,108 @@
+  # We don't _need_ `lib` here, but importing it wouldn't be detrimental to performance either because of Nix's `import` cache
+  inherit (builtins) attrNames elemAt length mapAttrs readDir;
+
+  # Warning: Do not make this configurable, otherwise the `import` evaluation cache won't work!
+  baseDirectory = ../by-name;
```
It shouldn't be in `lib` (yet) because it's only an internal standard for now. The suggested `isAttrs` test should take care of ensuring this, I'll do that.

roberth replied:
That's fair. A separate file may already be helpful for testing and achieve some of the same effect.

roberth replied:
`binaryMerge` could be added to `lib` before anything `callPackage`-specific.

infinisil replied:
Sounds good, though note that this will delay this PR further.

I'll go for an interface like this (name bikeshedding appreciated):
`mapMergeAttrs :: (Any -> Attrs) -> [ Any ] -> Attrs`

this should work with minimal performance loss

infinisil replied:
I opened https://github.com/NixOS/nixpkgs/pull/244187 for this

Though I changed the function to just be `lib.attrsets.mergeAttrsList :: [ Attrs ] -> Attrs`. I measured it, there's no performance loss with this.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-07-13-documentation-team-meeting-notes-63/30450/1

oxij reviewed (changes requested):
I wholeheartedly support the idea of this (given I wanted it since #50643), but I still think that `pkgs/top-level/named-hierarchy-files.nix` should be autogenerated via an external script (which then can be run in CI to check that the working tree contains the correctly generated file).

As is, this PR introduces more nix-evaluation to simply get a package list, which seems counter productive given that `nix-env -qaP --out-path` takes >24G of memory already.

roberth commented:
Autogenerated code would additionally retain a copy of the AST in memory, whereas this solution only retains the attrset.

oxij commented:
@roberth
> Autogenerated code would additionally retain a copy of the AST in memory, whereas this solution only retains the attrset.
Even assuming nix does this (keeps the source file in memory forever) for some reason (which AFAIK it does not), you are not actually winning anything, because you will generate all those `baseDirectory + "/${shard}/${name}/package.nix"` paths anyway. Even if readdir(2) and merging of attrsets costed nothing, this would still be less efficient than a plain generated  `.nix` file, since this code will store most of those paths as closures of  `(+) baseDirectory "/${shard}/${name}/package.nix"`.

roberth commented:
> Even assuming nix does this (keeps the source file in memory forever)

Please don't write about false assumptions, and let's stop speculating. I don't think the difference would be significant compared to the human cost of maintaining the list, at least for the time being.

If you can come up with an objectively faster and/or more memory efficient method, we can consider that separately. For now the goal is to implement the RFC, and ~10 ms or ~4 MB are not blocking issues.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-07-17-documentation-team-meeting-notes-64/30653/1

oxij commented:
I'm not saying the list should be maintained by hand, I'm saying it should be autogenerated with a script, for which I've shown you the math.

But alright, whatever, the migration to the new file hierarchy is the important step, the 244x slowdown would be easy to fix afterwards.

piegamesde commented:
> I'm not saying the list should be maintained by hand, I'm saying it should be autogenerated with a script

You've been understood, but this does not negate the point about additional maintenance cost of your proposal compared to this one.

> , for which I've shown you the math.

You've done some napkin calculations based on assumptions which are hard to verify from the outside. Besides, I wouldn't trust any numbers without benchmarks anyways.

The point is, as long as this change does not produce significant performance regressions, it does not really matter.

P.S.: please keep your sarcasm for yourself, I liked being in a good mood.

infinisil commented:
@piegamesde The estimate by @oxij is actually very accurate, it's close to the benchmark results I got (see [here](https://github.com/NixOS/nixpkgs/pull/237439/files#diff-c2d9776877bdbfe99d12936253c6457472358e40f484120d5546b841ca8c6070R97) (I should add to this PR that I performed that benchmark with all top-level attributes added to `pkgs/by-name`).

But yeah agreed with @roberth and @piegamesde here, the cost of having to maintain an auto-generated file by far outweighs the slightly decreased performance. But I do appreciate @oxij's general concern for performance, we should definitely take a closer look at performance for the second PR, which will do the big migration: https://github.com/NixOS/nixpkgs/pull/211832.

However I have effectively 0 concerns about performance, it's a one-time computation that takes 7MB and 15ms, which is insignificant compared to the resources most Nix evaluations require. But yes, let's optimise where we can, small savings add up. I think the best solution to improve performance would be to add a Nix builtin though, then we can have the cake and eat it too :)

oxij commented:
@piegamesde 

> You've been understood, but this does not negate the point about additional maintenance cost of your proposal compared to this one.

True.

> You've done some napkin calculations based on assumptions which are hard to verify from the outside. Besides, I wouldn't trust any numbers without benchmarks anyways.

I and @danbst did benchmark #50643, and @danbst did even benchmark various hypothetical implementations, see https://github.com/NixOS/nixpkgs/pull/50643#issuecomment-457202241

To quote from there:

> So, the result for 40000 packages:
> 
>     * original runtime is 0.168197s, oxij's list is 0.144745s, index version is 0.08515s
> 
>     * original has 96 bytes per package, oxij's list has 416, index has 96
> 
>     * original has 1,7M on disk, oxij's has 927K, index has 498K

And I'm saying we should adopt something like the "index" version from there here. The difference in real terms is significant even for such a small change as offloading `callPackage` calls. This doesn't. And actually _adds_ computation compared to `all-packages.nix`.

But this is easier to maintain, true.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/could-there-ever-be-some-kind-of-config-d-or-imports-globbing-for-easier-programmatic-editing/30668/4

infinisil commented:
This is a very important change, but I'd like to take a little detour by first cleaning up the general contributing documentation with https://github.com/NixOS/nixpkgs/pull/245243. In recent weeks it really sunk in how bad Nixpkgs documentation is, and because I don't want to shoehorn this updated documentation into this mess, I think the time is better spent improving the docs overall. The linked PR is still an early draft, but feel free to give general feedback.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-07-17-documentation-team-meeting-notes-64/30653/2

infinisil commented on pkgs/top-level/named-hierarchy-files.nix:
I now call this `read-by-name.nix`, since it can also take an argument now (the import caching is done by `stage.nix` instead), this allows reusing it for the tests

infinisil commented:
I have split this PR up once more into https://github.com/NixOS/nixpkgs/pull/250885, which just adds the checking tool, without enabling it in Nixpkgs.

infinisil commented:
The above split-off and this PR is now scheduled to be reviewed and merged with the [Nixpkgs Architecture Team](https://nixos.org/community/teams/nixpkgs-architecture.html) and anybody else that wants to join in https://meet.jit.si/nixpkgs-architecture on these times:
- 2023-08-29 15:30 CEST: https://github.com/NixOS/nixpkgs/pull/250885
- 2023-09-05 15:30 CEST: (this PR)

amjoseph-nixpkgs reviewed (commented):
From [this comment](https://github.com/NixOS/nixpkgs/pull/250885#issue-1862467201):

> This allows [RFC 140] Simple package paths, part 1b: Enabling the directory structure #237439 to then just fetch the tool from the nixpkgs-unstable channel, without having to do any compilation, allowing it to check all PR's without any delay, even ones that would cause mass rebuilds.

When I first read this, it sounded like CI would always use the latest (as of the moment when CI is run) Hydra binary for `checkReferences` from the `nixpkgs-unstable` channel.

This would be bad, because if a bug in `checkReferences` were preventing CI from passing, we would not be able to fix the bug in the same PR as the change it was incorrectly blocking.  In fact, we would need to not only submit two PRs, but wait for a channel-advance to occur between them.  That would be extremely annoying.

However after briefly skimming this PR, I can't find anything that obviously looks like "fetch the tool from the nixpkgs-unstable channel".

Could you clarify which of these is the case?

1. `checkReferences` is always fetched from Hydra, even if Hydra is out-of-date from the perspective of the PR upon which CI is being run
2. `checkReferences` is fetched using Hydra as a cache, just like every other package

If 1 is correct then that is a serious problem.

If 2 is correct then it's just a wording issue.

infinisil commented:
@amjoseph-nixpkgs It's not just `checkReferences` anymore, but rather the `nixpkgs-check-by-name` CLI tool. I just wrote some contributor docs in that PR [here](https://github.com/NixOS/nixpkgs/blob/f41d4792843cffd71a6de1ca1fd25009ba628df4/pkgs/test/by-name/README.md) which contain some explanation and justification.

> When I first read this, it sounded like CI would always use the latest (as of the moment when CI is run) Hydra binary for `checkReferences` from the `nixpkgs-unstable` channel.

Yes, that's right. CI should fetch the `nixpkgs-unstable` channel, build (but actually just substitute) `-A tests.byName` from cache.nixos.org, then run it on the PR's tree.

> This would be bad, because if a bug in `checkReferences` were preventing CI from passing, we would not be able to fix the bug in the same PR as the change it was incorrectly blocking. In fact, we would need to not only submit two PRs, but wait for a channel-advance to occur between them. That would be extremely annoying.

I guess I haven't mentioned how I intend to implement the CI check. It should be a GitHub Action that does the above-described process. It being a GitHub Action will then allow unblocking CI immediately by just removing or turning off the respective `.github/workflows` file, so this shouldn't be a big concern. Also note that all Nixpkgs committers can turn off or rerun GitHub Actions workflows [here](https://github.com/NixOS/nixpkgs/actions).

> However after briefly skimming this PR, I can't find anything that obviously looks like "fetch the tool from the nixpkgs-unstable channel".

This PR is not up-to-date with the latest intentions and that's why I'm still leaving it as a draft. I will mark the PR as ready to review once it's up-to-date and mergeable. I will work hard to get to that point before the scheduled date.

amjoseph-nixpkgs commented:
> This PR is not up-to-date with the latest intentions and that's why I'm still leaving it as a draft. I will mark the PR as ready to review once it's up-to-date and mergeable. I will work hard to get to that point before the scheduled date.

Okay; when you do update this with the mechanism described above would you please ping me?

Disabling (parts of) CI seems like quite a sledgehammer.  Not excited about that being the only alternative to "submit a separate PR, wait for it to merge, wait for the channel to advance."  It would be much better if the cache were used only *as a cache* rather than as a source of truth.

Maybe there's some fundamental reason why it can't work that way.  I'll wait for the undraftified PR to see if that's the case.  I just don't want to make the "ofborg problem" (our CI tool lives in a different tree than the code it does CI upon) worse than it already is.

amjoseph-nixpkgs commented on pkgs/test/by-name/check-references/README.md:
```diff
@@ -0,0 +1,3 @@
+Usage: nix-reference-check [FILES]
```
```suggestion
Usage: check-references [DIR]
```

At least when I built it, the binary that got built was called `check-references`.

Also, it insists on being passed a directory; you can't pass it just a single `.nix` file.

infinisil replied:
This is outdated, the check is now in https://github.com/NixOS/nixpkgs/pull/250885, called `nixpkgs-check-by-name <NIXPKGS>`, and there's no support for files (it takes a Nixpkgs path now).

amjoseph-nixpkgs commented on pkgs/test/by-name/check-references/README.md:
```diff
@@ -0,0 +1,3 @@
+Usage: nix-reference-check [FILES]
+
+Checks that all given Nix files don't transitively and statically refer to files outside their own directory
```
```suggestion
Checks that all given Nix files don't statically refer to files outside their own directory
```

amjoseph-nixpkgs commented on pkgs/test/by-name/check-references/README.md:
```diff
@@ -0,0 +1,3 @@
+Usage: nix-reference-check [FILES]
+
+Checks that all given Nix files don't transitively and statically refer to files outside their own directory
```
```suggestion
Checks that all given Nix files don't transitively and statically refer to files outside the directory which contains them.

Dynamic references, produced by using `+` to concatenate a path with a string, are not checked.  For example, this tool will not prevent:

  import (./. + "/../../../default.nix")
```

amjoseph-nixpkgs replied:
I tried this test case and `check-references` allowed it.  IIUC this is intended, since the tool only does what can be done without having to evaluate anything.

infinisil replied:
Oh yeah that's a good point. I'm not sure to what extent these exact semantics should be specified though. It's not really important for contributors to know the exact rules, and if they're interested, they could check the code or try it out. And I'd hope that reviewers catch people doing such workarounds.

amjoseph-nixpkgs commented on pkgs/test/by-name/check-references/src/main.rs:
```diff
@@ -0,0 +1,160 @@
+        return Ok(());
+    }
+
+    let rel_file = file.strip_prefix(&pkg_dir).unwrap();
```
I tried running this tool on `pkgs/development/compilers/gcc/` at current nixpkgs `master`, commit e9b86d7b8b0b68bcb4dcf6770ed2e85652b0594e and got this error:

```
$ RUST_BACKTRACE=1 ~/result/bin/check-references pkgs/development/compilers/gcc/
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: StripPrefixError(())', src/main.rs:80:48
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: check_references::check_path
   4: check_references::check_path
   5: check_references::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

infinisil replied:
Should not be a problem anymore in the [split-off tooling PR](https://github.com/NixOS/nixpkgs/pull/250885).

amjoseph-nixpkgs commented on pkgs/test/by-name/check-references/Cargo.toml:
```diff
@@ -0,0 +1,8 @@
+[package]
+name = "check-references"
```
This tool doesn't seem to handle subdirectories (below the directory containing `package.nix`) the way the RFC intends.

For example, I tried running it on `pkgs/development/compilers/gcc` has a subdirectory `common` which contains `.nix` files which refer to `../patches/` at commit 86d0c37bba6b7f8ca20ebd834e793d05be51e40e (this PR) and got a lot of errors:

```
- pkgs/development/compilers/gcc/10/default.nix: File  at line 57 contains the path expression "../fix-bug-80431.patch" which may point outside the directory of that package.
- pkgs/development/compilers/gcc/10/default.nix: File  at line 58 contains the path expression "../11/fix-struct-redefinition-on-glibc-2.36.patch" which may point outside the directory of that package.
- pkgs/development/compilers/gcc/10/default.nix: File  at line 59 contains the path expression "../libstdc++-target.patch" which may point outside the directory of that package.
- pkgs/development/compilers/gcc/10/default.nix: File  at line 60 contains the path expression "../no-sys-dirs.patch" which may point outside the directory of that package.
- pkgs/development/compilers/gcc/10/default.nix: File  at line 61 contains the path expression "../no-sys-dirs-riscv.patch" which may point outside the directory of that package.
...
```

My understanding is that these sorts of references are allowed by the RFC, because these references remain within the `gcc` package directory (which will soon be `pkgs/by-name/gc/gcc/`).

amjoseph-nixpkgs replied:
I think `check-references` needs to maintain a counter of how many levels below the starting directory it is, and allow up to that many `../`'s.

infinisil replied:
Should not be a problem anymore in the [updated tooling PR](https://github.com/NixOS/nixpkgs/pull/250885).

amjoseph-nixpkgs reviewed (commented):
Nifty tool!  I tried it out but ran into a few problems.

infinisil commented:
@amjoseph-nixpkgs As I just said earlier, this PR is not ready yet, in particular I split off the entire checking code into https://github.com/NixOS/nixpkgs/pull/250885, the code here is outdated and won't be included like this! That _other PR_ is _almost_ ready though (only a bit cleaner code is missing, I'll unmark it as draft very soon), feel free to review that one already!

infinisil commented:
The other PR (a prerequisite for this one) is now ready to be reviewed! https://github.com/NixOS/nixpkgs/pull/250885

infinisil committed:
pkgs/by-name: Introduce

This introduces the `pkgs/by-name` directory as proposed by RFC 140.
Included are:
- The implementation to add packages defined in that directory to the
  top-level package scope
- Contributer documentation on how to add packages to it
- A GitHub Actions workflow to check the structure of it on all PRs

infinisil committed:
pkgs/README.md: Update to mention pkgs/by-name

infinisil committed:
hello: Move to pkgs/by-name

infinisil commented:
After the `nixpkgs-check-by-name` tool was [successfully merged](https://github.com/NixOS/nixpkgs/pull/250885) this week, including a [channel fix](https://github.com/NixOS/nixpkgs/pull/252210) and a [backport](https://github.com/NixOS/nixpkgs/pull/252154), I went on to readying up this PR:
- Introduce the GitHub Actions workflow to run the tool on every PR (see the new "Check pkgs/by-name" in the PR checks!)
- Update the documentation according to previous feedback
- Add more comments and tidy up everything in general

I can confirm that this now works very well and I consider this ready to be merged!

As with the other PR, we have a meeting scheduled for next Tuesday, 2023-09-05 15:30 CEST in https://meet.jit.si/nixpkgs-architecture to do a final review and merge (if everything looks good) with the Nixpkgs Architecture Team and anybody else that wants to join :smile:

So please review this until Tuesday and we'll have it merged for Nixcon!

roberth commented on pkgs/top-level/by-name-overlay.nix:
```diff
@@ -0,0 +1,46 @@
+self: super:
+mapAttrs (name: file:
+  self.callPackage file { }
+) packageFiles
```
Did you try this? It might release a significant number of thunks, including their closures, including the `readDir` attrsets and such.

```suggestion
) (builtins.deepSeq packageFiles packageFiles)
```

The cost of the computation may or may not pay off thanks/despite reduced gc times later.
As for memory I would expect a reduction, but I haven't tried it. The way to measure it is with the final heap size (`NIX_SHOW_STATS=1`), but unfortunately that measure is rounded and noisy. I think Nix needs to do a full GC before reporting stats...

infinisil replied:
I didn't try this, but I now added a comment suggesting this for a potential future optimisation: [diff](https://github.com/NixOS/nixpkgs/compare/4d49c6b94cc40b4ae1afab8eb33c4f78ef9122cd..ebea45378c8d4bd2efdee343c3422509f64ca7b3#diff-caaafdd48fde2a02cd655ca77794b57e1f72f1edce04b239ab72f212ba9d1608)

YorikSar commented on pkgs/by-name/README.md:
````diff
@@ -0,0 +1,107 @@
+But if the package needs e.g. `pkgs.libbar_2` instead, you can change the default `libbar` attribute argument by adding a definition as follows to [`pkgs/top-level/all-packages.nix`](../top-level/all-packages.nix):
+```nix
+libfoo = callPackage ../by-name/so/somePackage/package.nix {
+  libbar = libbar_2;
````
Since `libbar_2` would be imported in `by-name-overlay.nix`, it would be

```suggestion
  libbar = pkgs.libbar_2;
```

right?

YorikSar replied:
Oh, wait, it's somePackage that's in by-name, not `libbar`. And we're already not using `pkgs.` for versioned packages.

infinisil replied:
There's effectively no difference whether something is defined using `pkgs/by-name` or anywhere else. And also in `all-packages.nix`, there's a `with pkgs;` at the top, so `libbar_2 == pkgs.libbar_2` :)

infinisil commented:
@roberth Thanks for the suggestions, I reworded them a bit, but mostly applied them ([diff](https://github.com/NixOS/nixpkgs/compare/4d49c6b94cc40b4ae1afab8eb33c4f78ef9122cd..ebea45378c8d4bd2efdee343c3422509f64ca7b3))

infinisil commented on pkgs/by-name/README.md:
````diff
@@ -0,0 +1,107 @@
+};
+```
+
+This allows maintaining the same `.override { libbar = ...; }` interface for package consumers.
````
Reworded a bit, but that's a good point yes :smile: [diff](https://github.com/NixOS/nixpkgs/compare/4d49c6b94cc40b4ae1afab8eb33c4f78ef9122cd..ebea45378c8d4bd2efdee343c3422509f64ca7b3#diff-a8563497765f3c15777f4e7ebdfa64a8d7ac5fa696f5e1b141bd38545d242a46L61-L69)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-09-05-nixpkgs-architecture-team-meeting-43/32635/1

phaer commented on pkgs/by-name/README.md:
````diff
@@ -0,0 +1,99 @@
+```
+
+But the package might need `pkgs.libbar_2` instead.
+While the function be changed to take `libbar_2` directly as an argument,
````
```suggestion
While the function could be changed to take `libbar_2` directly as an argument,
```

infinisil reviewed (commented):
As planned we looked at this together in today's meeting with the Nixpkgs Architecture Team (including @aakropotkin, @roberth, @DavHau, @phaer), and @fgaz, @Quantenzitrone and @YorikSar.

Everything was looked at and minor documentation updates were done. Nobody had any complaints about this PR, so we'll go ahead with merging it! :rocket:

infinisil commented:
Backport PR, let's make sure this one doesn't have any problems before merging it though: https://github.com/NixOS/nixpkgs/pull/253442

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nix-init-generate-nix-packages-from-urls-with-hash-prefetching-dependency-inference-license-detection-and-more/25035/18

doronbehar commented:
I'm getting the following error from CI on a [new PR I opened](https://github.com/NixOS/nixpkgs/pull/255902), from the `check-by-name.yml` workflow:

```
  Error: fatal: couldn't find remote ref refs/pull/255902/merge
  The process '/usr/bin/git' failed with exit code 128
```

infinisil commented:
Thanks for the report! Opened https://github.com/NixOS/nixpkgs/issues/256756 to discuss this

SuperSandro2000 commented on pkgs/README.md:
````diff
@@ -49,22 +50,25 @@ Now that this is out of the way. To add a package to Nixpkgs:
    ```ShellSession
-   $ emacs pkgs/development/libraries/libfoo/default.nix
-   $ git add pkgs/development/libraries/libfoo/default.nix
+   $ emacs pkgs/by-name/so/some-package/package.nix
+   $ git add pkgs/by-name/so/some-package/package.nix
````
IMO this should just be default.nix to follow the previous pattern. The argument for naming it package.nix is rather weak and just creates confusion since there is no real reason to name it different.

infinisil replied:
This is part of the RFC so we can't just change that, and you already started a discussion on this in https://github.com/NixOS/rfcs/pull/140#discussion_r1367936597, let's keep it in one place

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/ci-will-soon-enforce-pkgs-by-name-for-new-packages/38098/1
